### PR TITLE
feat(new-client): on-demand revalidation

### PR DESCRIPTION
**Pull request by Nirajn2311**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

To quote a wise developer - "Closes nothing and almost certainly opens many things."

Currently the new client uses Incremental Static Regeneration to update cached static page(if the data has changed) every 10 seconds which in turn means fetching data from the server every 10 seconds. 

Next.js has released a **BETA** version of On-demand revalidation([link](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation-beta) to docs) which allows manually revalidation of a cache. This will help us as we can trigger the revalidation only when challenge data has been updated instead of periodically checking for data change thereby reducing the number of requests done to the server.

I've used `nodemon` to implement a restartable server, which is how I'm assuming the final server will work when data is changed. I've also added scripts for building and starting the next instance to test the implementation in production mode(static pages are always generated for each request in development mode and only act like they're intended in production mode).

I've implemented this feature only for the challenge list page and am yet to do it for individual challenge pages.

cc - @ojeytonwilliams

### CODE REVIEW

**Review by ojeytonwilliams — COMMENTED** [review]
> To quote a wise developer - "Closes nothing and almost certainly opens many things."

:100: 

Hey @Nirajn2311 this looks great and I know this is a draft, but "If you don't want a review, keep it on your machine" - someone, probably.

One thing I spotted when I ran `npm run build`:
```
[develop-curriculum-server] JSON Server is running
[develop-curriculum-server] ECONNREFUSED connect ECONNREFUSED 127.0.0.1:8000
```

File: web/src/pages/api/revalidate.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  // if (req.query.secret !== process.env.REVALIDATION_SECRET) {
+  //   return res.status(401).json({ message: 'Inavlid token' });
+  // }
+
+  try {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const staticPaths = await getStaticPaths();
+    console.log(staticPaths.paths);
+    await res.unstable_revalidate(staticPaths.paths[0] as string);
```
[the remark this answers is not in the document]
Comment (Nirajn2311) [reply web/src/pages/api/revalidate.ts:12]: If we want to revalidate more paths then we pass them one by one. I referred to this discussion when making this solution - https://github.com/vercel/next.js/discussions/34567

**Comment by Nirajn2311** [comment]
> One thing I spotted when I ran `npm run build`:

So the server calls the `revalidate` endpoint when it starts up. This was the only was I could think of(and get to work) wherein the server is up when the endpoint is pinged everytime there is a change in data.

**Comment by ojeytonwilliams** [comment]
I see, thanks. This behaviour is something we'd only want in development, so I can live with the error for now.

That said, could you control it with FREECODECAMP_NODE_ENV? I'd like to encode the 'dev-only' status early.

**Comment by Nirajn2311** [comment]
> could you control it with FREECODECAMP_NODE_ENV?

The client or the server?

**Comment by ojeytonwilliams** [comment]
> The client or the server?

The server.  I production we'll want something else (like pm2) to control restarts, not nodemon.

**Comment by raisedadead** [comment]
Hi, 

We are closing old issues/pull requests that we think are stale. If you still need help or are fixing this, please open a new issue/pull request.

Thanks for your understanding. 

Happy contributing.

**Pull request closed without merging**